### PR TITLE
Add military_status to user (fixes #373)

### LINKED ISSUE #373: Identify military spouses in backend admin interface
Not sure if this is a feature request or a bug.

# Bug Report
## What is the current behavior?
Our current user signup flow asks for information on whether someone is a military spouse, but this does not appear to be visible in the backend admin interface.

## What is the expected behavior?
Ability to search by military spouse.

### PR DESCRIPTION
**Pull request by hollomancer**
# Description of changes
Adds a column for `military_status`, which is currently not present on `users`. This will allow the information we request regarding a military spouse / active duty in signup flow to save to the DB.
# Issue Resolved
Part of fixing #373 (some frontend work needed)

_It's been a while for me. Please review._

### CODE REVIEW

**Comment by hollomancer** [comment]
Also fixes #374

File: db/migrate/20180711212702_add_militarystatus_to_user.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class AddMilitarystatusToUser < ActiveRecord::Migration[5.0]
+  def change
+    add_column :users, :military_status, :string
```
Comment (hpjaj) [inline db/migrate/20180711212702_add_militarystatus_to_user.rb]: As this is a `string`, is the plan to accept any `string`, or a specific list of strings for consistency?  Meaning a specified set of four statuses, for example.  I **highly** recommend the latter.  

This way, we can enforce data integrity, hang logic on the responses, do data analysis, filtering and searching, add rules for certain statuses, etc., etc.

If the answer is yes, then we'll need to do a few extra additions in this PR before shipping it.

Comment (hollomancer) [reply db/migrate/20180711212702_add_militarystatus_to_user.rb]: I'd just need to add the check in `/app/models/user.rb`, is that correct?

Comment (hpjaj) [reply db/migrate/20180711212702_add_militarystatus_to_user.rb]: @hollomancer - To establish the specific set of validated statuses, you'll want to:

1. Create a constant for each status [here](https://github.com/OperationCode/operationcode_backend/blob/master/app/models/user.rb#L22).  
2. Create a constant to represent an array of all of the statuses.  You'll need to include one for `nil`, to account for our current users, and future users, that will not have a status.  For example:

```ruby
CURRENT = 'current'
VETERAN = 'veteran'
SPOUSE = 'spouse'
MILITARY_STATUSES = [CURRENT, VETERAN, SPOUSE, nil]
```

3. In `models/user.rb`, Add an `inclusion` validation for the new `:military_status` attribute, for the `MILITARY_STATUSES`.  [Here is how to do that.](http://guides.rubyonrails.org/active_record_validations.html#inclusion) 
4. In [`admin/user.rb`](https://github.com/OperationCode/operationcode_backend/blob/master/app/admin/user.rb#L93), add the set of selections to ActiveAdmin dashboard **form** portion.  For example:

```ruby
f.input :military_status, as: :select, collection: User::MILITARY_STATUSES.map { |status| [status, status].compact }.reject(&:empty?), include_blank: true
```

5. Add test coverage to `user_test.rb` testing the new inclusion validation
6. In [apiary.apib](https://github.com/hollomancer/operationcode_backend/blob/c0d852e8481d229d520652c7051567ff5af4ed7c/apiary.apib#L1810), in the `military_status` definition, change that to something like:

```
Users military status, either: current, veteran, spouse, or blank
```

**Comment by dmarchante** [comment]
@hpjaj is this only waiting a review, or are there more changes to be made?

**Comment by hollomancer** [comment]
@dmarchante, this one's on me to finish validation appropriately.

File: test/models/user_test.rb [hunk 2]
```diff
@@ -1,6 +1,6 @@
 require 'test_helper'
 
-class UserTest < ActiveSupport::TestCase
+class UserTest < ActiveSupport::TestCase # rubocop:disable Metrics/ClassLength
```
Comment (hollomancer) [inline test/models/user_test.rb]: @hpjaj, should we break this up into smaller segments?

Comment (hpjaj) [reply test/models/user_test.rb]: No, let's keep all of the user model related test code in this file.  

As far as disabling RuboCop, you can:

- do it like this: http://www.mindonrails.com/notes/334-disabling-and-re-enabling-a-rubocop-rule-for-a-specific-set-of-code
- add an exclusion for the test file to `.rubocop.yml`

The latter is slightly preferable.

File: app/admin/user.rb [hunk 3]
```diff
@@ -105,13 +106,15 @@
...
       f.input :years_of_service
+      f.input :military_status, as: :select, collection: User::MILITARY_STATUSES.map { |status| [status, status].compact }.reject(&:empty?), include_blank: true
       f.input :pay_grade
       f.input :military_occupational_specialty
       f.input :github
       f.input :twitter
       f.input :linkedin
       f.input :employment_status
       f.input :education
+      f.input :military_status
```
Comment (hpjaj) [inline app/admin/user.rb]: You'll want to delete this input (we'll favor the new collection input on line 109).

File: test/models/user_test.rb [hunk 4]
```diff
@@ -260,4 +260,11 @@ def user_opts
     assert_equal data[:zip], userInfo[:zip]
     assert_equal '/profile', redirect
   end
+
+  test 'military status accepts valid states' do
+    assert create(:user, military_status: 'spouse')
```
Comment (hpjaj) [inline test/models/user_test.rb]: Can you pls change these three to pass the three new constants, to support our validation?  i.e.

```ruby
assert create(:user, military_status: User::SPOUSE).valid?
```

**Pull request merged**